### ISSUE TITLE
(NumPy)Printer does not handle contracted indices in Indexed

### ISSUE THREAD

**Issue by tBuLi** [sympy/sympy#15229]
I apologize if this is a known issue but I couldn't find it.

Consider the following (using sympy 1.2):
```python
A_mat = np.array([[1, 1, 1], [0, 2, 5], [2, 5, -1]])
y_vec = np.array([6, -4, 27])
x_vec = np.array([5., 3., -2.])

x, a = symbols('x, a', cls=IndexedBase)
i, j = symbols('i, j', range=len(x_vec), cls=Idx)
y=  a[i, j] * x[j]

from sympy.printing.pycode import NumPyPrinter as Printer
from sympy.utilities.lambdify import lambdastr
lstr = lambdastr([a, x, i, j], y, printer=Printer, dummify=False)
print(lstr)
```
prints 
```
lambda a,x,i,j: (a[i, j]*x[j])
```
This makes sense, but is not the desired result. I therefore wrote a function which simply replaces implicit sums by explicit sums, such that the result becomes `y=  Sum(a[i, j] * x[j], j)`. But then we get:
```python
y=  Sum(a[i, j] * x[j], j)
lstr = lambdastr([a, x, i], y, printer=Printer, dummify=False)
print(lstr)
```
prints 
```
lambda a,x,i: ((builtins.sum(a[i, j]*x[j] for j in range(0, 2+1))))
```
Better, but I'm a bit supriced by the `builtins.sum` when asking for a numpy implementation. For bigger tensors, this will not have good performance. Instead, I think the correct replacement is `numpy.tensordot`.

Therefore, instead of replacing the implicit sum by `Sum`, I now replace by the following new object:
```python
class IndexedMul(sympy.Mul):
    """
    Represent the multiplication of Indexed terms. Behaves exactly like a `Mul`,
    except that it is printed into a tensordot over the relevant indices.
    """
    def __new__(cls, *args, contracted, **options):
        return super(IndexedMul, cls).__new__(cls, *args, **options)

    def __init__(self, *args, contracted, **options):
        self.contracted = contracted
        return super(IndexedMul, self).__init__()

    def _sympystr(self, printer, *args, **kwargs):
        """
        This object should be printed as a tensordot.
        """
        basestr = 'tensordot({0}, {1}, axes={2})[{3}]'
        # give the Indexed to be dotted a simpler name
        a = self.args[0]
        b = self.args[1]
        # Identify the axes to be contracted
        summed_a = [axis for axis, index in enumerate(a.indices) if index in self.contracted]
        summed_b = [axis for axis, index in enumerate(b.indices) if index in self.contracted]
        sum_axes = (summed_a, summed_b)
        # We need the uncontracted as well, as these become indices on the resulting tensor.
        uncontracted = [index for index in a.indices + b.indices if index not in self.contracted]
        uncontracted_str = ','.join(str(i) for i in uncontracted)
        # build the result and doprint
        funcstr = basestr.format(a.base, b.base, sum_axes, uncontracted_str)
        return printer.doprint(funcstr)

    _numpycode = _sympystr
```
In `sympy 1.1.1` this works as desired, and prints: (In sympy 1.2 this raises a `TypeError`, but that is not the main story here so I'll pretend that didn't happen.)
```
lambda a,x,i: (tensordot(x, a, axes=([0], [1]))[i])
```
Also, it gives the correct result when evaluated:
```
>>> func = lambdify([a, x, i], implicit_to_explicit_sums(model_dict[y[i]]))
>>> print(func(a=A_mat, x=x_vec, i=...))
[ 6. -4. 27.]
```
I therefore think that the printing of `Mul` could be made to have this behavior for expressions containing `Indexed` contractions. This will greatly enhance performance.

If people agree and can point me in the right direction, I wouldn't mind making a PR out of this, but I'm a `sympy` novice so I would need adult supervision.

(p.s. I know that in this code there are some things I didn't consider yet, it's a first draft. Not least of which what happens when more than two `Indexed` are multiplied.)

**Comment by Upabjojr** [comment]
Vaguely reminds of an old PR: https://github.com/sympy/sympy/pull/9314

**Comment by Upabjojr** [comment]
I would avoid creating new classes. I have already objected in https://github.com/sympy/sympy/pull/9314 to the introduction of `einsteinsum`. I believe we should just use the existing `Sum` class.

**Comment by tBuLi** [comment]
I see your point, but don't you think it might be nice to use the more efficient dot products numpy offers?

**Comment by Upabjojr** [comment]
The NumPy code printer needs to be fixes to recognize dot products.

By the way, dot products make sense if you use the indexless notation for vectors. If indices are explicitly represented, there is no need for them.

The matrix expression module supports dot products as objects on their own.

**Comment by tBuLi** [comment]
> By the way, dot products make sense if you use the indexless notation for vectors. If indices are explicitly represented, there is no need for them.

I agree, but for my research the indexed module has far more potential than normal matrices and vectors. I work in lattice QCD, where we try to numerically simulate the particles from the standard model. There I work with many different kinds of indices, Lorentz, Color, spacetime coordinate to name a few. I could write that as a Matrix in a vector at every spacetime point or something like that, but these Indexed objects mean that I can write code that is very close to what we write by hand:
```python
U, g = symbols('U, g', cls=IndexedBase)  # Gauge field configuration U, Gauge transformation g
i, j = symbols('i, j', cls=Idx, range=3)  # SU(3) matrix indices
st = symbols('x, y, z, t', cls=Idx)  # spacetime coordinate indices
x, y, z, t = st

# Find the g which maximizes this trace for a given dataset for U
model = g[i, j, *st] * U[i, j, *st]

# g[i,j] is an SU(3) matrix at every point in spacetime. This means we want det(g[i,j]) == 1
# which can be expressed with a Levi_Civita as a contraction: 
# https://en.wikipedia.org/wiki/Levi-Civita_symbol#Determinants
constraints = [LeviCivita(i, j, k) * g[1, i, *st] * g[2, i, *st] * g[3, i, *st]]

fit = Fit(model, U=Udata, constraints=constraints)
fit_result = fit.execute()
```
where `Fit` comes from a fitting package I'm developing in parallel. Tell me that wouldn't be cool ;). But before I can do that, dot product definitely should work ;). So I'm invested in helping develop this feature further.

**Comment by Upabjojr** [comment]
What about extending the support of tensorcontraction( ) to IndexedBase?

**Comment by Upabjojr** [comment]
I still recommend using Sum instead of the dot product.

**Comment by Upabjojr** [comment]
Otherwise a function which finds the repeated indices and adds Sum objects. Or you can try keeping a separate set with the indices to be summed over.

I mean, there are lots of ways you can proceed.

**Comment by tBuLi** [comment]
> I still recommend using Sum instead of the dot product.
No argument from me there.

> Otherwise a function which finds the repeated indices and adds Sum objects.
Locally I already do exactly that. The main problem for me now is that such `Sum` are turned into a `builtin.sum` by the printer, and not recognized as a dot product.

**Comment by Upabjojr** [comment]
So it's a printer problem. Which printer are you using?

By the way... I am reviewing the printers for SymPy right now, so tell me if you have issues.

**Comment by tBuLi** [comment]
This particular problem happens with the NumpyPrinter, I haven't tested it for other printers.

And now that you ask, I have also had this problem before: #14911. My fix for that has been merged, but the conceptual discussion I had there with @asmeurer might be relevant if you are reviewing the printers. Thanks for your work on this btw.

**Comment by Upabjojr** [comment]
Yes, the printer for `Sum` is not so easy, because it can be printed into different things. In NumPy it could be represented by `tensordot` if it's an array contraction, but in many other cases it may be something else.

**Comment by Upabjojr** [comment]
I have added https://github.com/sympy/sympy/pull/15427

It is not finished.
